Added Redfish Management Clean steps

DhuldevValekar3 opened the pull request:
Hi Chris,

I have added below redfish management clean steps :
1. clear_job_queue
2. reset_idrac
3. known_good_state

Please have a look.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -308,7 +331,143 @@ class DracRedfishManagement(redfish_management.RedfishManagement):
     specific incompatibilities and introduction of vendor value added
     should be implemented by this class.
```
Can you remove the 4 lines above since it is no longer true with the addition of the code below?

DhuldevValekar3 replied:
Yes, removed

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -308,7 +331,143 @@ class DracRedfishManagement(redfish_management.RedfishManagement):
+                delete_response = requests.delete(
+                    job_url, headers=_HEADERS, verify=False,
+                    auth=(self.redfish_username, self.redfish_password))
+                LOG.info(
```
I think this logging statement needs to be unindented 4 spaces so that it isn't repeated once for each job.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -308,7 +331,143 @@ class DracRedfishManagement(redfish_management.RedfishManagement):
+            for job_id in job_list:
+                job_id = job_id.strip("'")
+                job_url = '%s/%s' % (jobs_url, job_id)
+                delete_response = requests.delete(
```
It looks like the job queue can be cleared in 1 call, without having to loop over the job IDs by passing in "JID_CLEARALL" for the job ID.  See: https://github.com/dell/iDRAC-Redfish-Scripting/blob/cbff04525e97e29b2714f9d91b6fbacc335bbd46/Redfish%20Python/GetDeleteJobQueueREDFISH.py#L31
and:
https://github.com/dell/iDRAC-Redfish-Scripting/blob/cbff04525e97e29b2714f9d91b6fbacc335bbd46/Redfish%20Python/GetDeleteJobQueueREDFISH.py#L84
Also, we need to check the http error code returned and raise a DracOperationError if the clear failed.

DhuldevValekar3 replied:
Addressed by passing job ID's as "JID_CLEARALL".

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -84,6 +89,24 @@
+_HEADERS = {'content-type': 'application/json'}
+
+# idrac remote service status retries constant
+_READY_RETRIES = 24
```
We should make the above constant have the same value as DEFAULT_IDRAC_IS_READY_RETRIES in python-dracclient:
https://github.com/openstack/python-dracclient/blob/master/dracclient/constants.py#L15

DhuldevValekar3 replied:
Addressed. Used this constant values same DEFAULT_IDRAC_IS_READY_RETRIES.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -308,7 +331,143 @@ class DracRedfishManagement(redfish_management.RedfishManagement):
+        if task.node.driver_internal_info.get("clean_steps"):
+            if task.node.driver_internal_info.get("clean_steps")[0].get(
+                    'step') in _CLEAR_JOBS_CLEAN_STEPS:
+                self.clear_job_queue(task)
```
We really shouldn't clear the entire job queue in this situation.  The user could be trying to run the reset iDRAC cleaning step, and they wouldn't expect the entire job queue to be cleared as a result of that.  We should only delete the pending jobs, like we do here:
https://github.com/openstack/ironic/blob/master/ironic/drivers/modules/drac/management.py#L207

DhuldevValekar3 replied:
Addressed, This will execute to delete only unfinished jobs in case of "clear_job_queue" and "known_good_state" clean steps.
If the clean step is only "reset_idrac" then this will not execute, as user wants to reset idrac only.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -308,7 +331,143 @@ class DracRedfishManagement(redfish_management.RedfishManagement):
+                reset_idrac_url,
+                data=json.dumps(payload), headers=_HEADERS, verify=False,
+                auth=(self.redfish_username, self.redfish_password))
+
```
We need to check the http error code returned and raise an exception on failure.

DhuldevValekar3 replied:
Addressed, moved low level redfish code to sushy-oem-idrac repo and added check for returned response code value.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -821,3 +825,85 @@ def test_clear_job_queue(self, mock_get_drac_client):
+                                  shared=False) as task:
+            task.node.driver_info['redfish_address'] = '1.2.3.4'
+            return_value = task.driver.management.clear_job_queue(task)
+            fake_system.clear_job_queue(task)
```
I don't think the line above does anything because you're just calling clear_job_queue on a mock.  It should probably be removed.

DhuldevValekar3 replied:
Addressed.

DhuldevValekar3 commented:
> Hey Dhuldev, I have a few comments inline in the code. There are a couple things to note:
> 
> 1. Whenever we make a request to the iDRAC, we need to check the http error code in the response and raise an exception if a failure was returned.
> 2. The low level Redfish code needs to be moved to https://opendev.org/x/sushy-oem-idrac.  This is the code that deals with requests and URIs.  This repo is meant to allow vendors to make extensions to sushy, and a fair amount of the code should be there instead of in ironic proper.  Moving the code there would make it so that this could could just use sushy instead of making redfish calls directly.  It will also change the implementation of the unit tests so that they look more like the ironic redfish driver unit test here:
>    https://github.com/openstack/ironic/tree/master/ironic/tests/unit/drivers/modules/redfish
> 
> See Ricks comments on this patch for more detail:
> https://review.opendev.org/#/c/693328/

Hi Chris,
Thanks for code review, will start on addressing review comments by considering sushy-oem-idrac repo.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -84,6 +88,14 @@
 _CLEAR_JOBS_CLEAN_STEPS = ['clear_job_queue', 'known_good_state']
 
+# System Management Constant
+_SERVICE_ROOT = '/redfish/v1/Managers/'
```
I think _SERVICE_ROOT can be remove.  Doesn't look like it's used.

DhuldevValekar3 replied:
Yes, It is unused constant value, Removed it.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+                LOG.debug("Sushy OEM extension Python package "
+                          "sushy-oem-idrac failed to clear job queue"
+                          "for %(node)s, Error : %(error)s" %
+                          {'node':task.node.uuid, 'error':e})
```
in this exception handler, I think you need to:
raise exception.RedfishError(error=error_msg)
otherwise, I think delete_job_response will be None, and the line below will cause a null pointer exception.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+                LOG.debug("Sushy OEM extension Python package "
+                          "sushy-oem-idrac failed to reset idrac"
+                          "for %(node)s, Error : %(error)s" %
+                          {'node':task.node.uuid, 'error':e})
```
Here too

DhuldevValekar3 replied:
Addressed, raised an exception on failure.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+        else:
+            LOG.error("Failed to clear job queue, node : %(node)s " %
+                     {'node': task.node.uuid})
+        return delete_job_response
```
Cleaning steps should either return states.CLEANWAIT or None.  See:
https://github.com/openstack/ironic/blob/master/ironic/drivers/base.py#L307
Since this cleaning step is synchronous, I think it should return None.

DhuldevValekar3 replied:
Addressed.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+        else:
+            LOG.error("Failed to reset idrac for %(node)s" %
+                    {'node': task.node.uuid})
+        return reset_job_response
```
This synchronous cleaning step should return None.

DhuldevValekar3 replied:
Done.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+                    {'node': task.node.uuid})
+            LOG.debug("iDRAC was reset, waiting for return to operational state")
+            redfish_utils.wait_for_host(redfish_node_ip)
+            LOG.info("The iDRAC has become pingable")
```
Can you combine the logging statements above and below into 1 logging statement?  Also, whenever we log something, we should include some identifying information about the object we're logging about.  In this case, we should include the node UUID in the logging statements about the node.
These should probably be at the debug level too.

DhuldevValekar3 replied:
Addressed, Combined multiple line logs into single log.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+        redfish_node_ip = re.search(r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}',
+                                    url).group()
+        if reset_job_response.status_code == _RESET_JOB_RESPONSE_CODE:
+            LOG.info("idrac reset success for node %(node)s via OEM" %
```
Can you combine the logging statements above and below into 1 logging statement?

DhuldevValekar3 replied:
Done, Combined logs into single log.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+                LOG.error(error_msg)
+                raise exception.RedfishError(error=error_msg)
+            try:
+                delete_job_response = oem_manager.clear_job_queue(
```
I think this should be calling clear_job_queue() in this class, not the one in OEM

DhuldevValekar3 replied:
Done.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+            try:
+                delete_job_response = oem_manager.clear_job_queue(
+                                                job_ids=['JID_CLEARALL'])
+                reset_job_response = oem_manager.reset_idrac(
```
We need to call reset_idrac() in this class so that it waits until the iDRAC is reset before returning.

DhuldevValekar3 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+
+    @METRICS.timer('DracRedfishManagement.reset_idrac')
+    @base.clean_step(priority=0)
+    def reset_idrac(self, task):
```
I think we need to split this into 2 methods: reset_idrac(), which returns None on success, and _reset_idrac() which returns the results.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+                    {'node': task.node.uuid})
+        else:
+            LOG.error("Failed to clear job queue, node : %(node)s " %
+                     {'node': task.node.uuid})
```
The cleaning step should raise an exception when an error occurs.  This is the only way we have to tell the conductor that there was a failure.

DhuldevValekar3 replied:
Addressed, raised an exception on failure.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
-    pass
+    @METRICS.timer('DracRedfishManagement.clear_job_queue')
+    @base.clean_step(priority=0)
+    def clear_job_queue(self, task):
```
I think we need to split this into 2 methods: clear_job_queue(), which returns None on success, and _clear_job_queue() which returns the results.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+                        retry_delay=IDRAC_IS_READY_RETRY_DELAY_SEC)
+        else:
+            LOG.error("Failed to reset idrac for %(node)s" %
+                    {'node': task.node.uuid})
```
We should raise an exception here so Ironic knows cleaning failed.

DhuldevValekar3 replied:
Addressed, raised an exception on failure.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: DracOperationError on an error from python-dracclient.
+        """
```
I think almost all of the code in this method could be remove and replaced with 2 lines that call into reset_idrac and clear_job_queue.  We should call reset_idrac first since that may get it into a state where clear_job_queue succeeds where otherwise it would have failed.

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+    else:
+        ping_type = "not pingable"
+
+    LOG.info("Waiting for the iDRAC to become %s", ping_type)
```
Need to include the IP of the iDRAC in this logging statement

DhuldevValekar3 replied:
Done, Logged iDRAC IP address while logging.

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+        else:
+            response_count = 0
+            if alive:
+                LOG.debug("The iDRAC is still not pingable")
```
Need to include the iDRAC IP in this logging statement

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+            if alive:
+                LOG.debug("The iDRAC is still not pingable")
+            else:
+                LOG.debug("The iDRAC is still pingable")
```
Need to include the iDRAC IP in this logging statement

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+                      ping_type,
+                      response_count)
+            if response_count == ping_count:
+                LOG.debug("Reached specified ping count")
```
Need to include the iDRAC IP in this logging statement

DhuldevValekar3 replied:
Done, Logged iDRAC IP address while logging.

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+    return state_reached
+
+def wait_for_host(redfish_node_ip):
+    LOG.debug("iDRAC was reset, waiting for return to operational state")
```
Need to include the iDRAC IP in this logging statement

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+    if retries == 0:
+        err_msg = "Timed out waiting for the iDRAC to become ready after reset"
+        LOG.error(err_msg)
+        raise
```
We should raise a DracOperationError here with the err_msg

DhuldevValekar3 replied:
Addressed, raised DracOperationError exception with err_msg

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+
+    LOG.info("The iDRAC has become pingable")
+
+def wait_until_idrac_is_ready(oem_manager, retries=None, retry_delay=None):
```
I think retries should be defaulted to IDRAC_IS_READY_RETRIES and retry_delay should be defaulted to IDRAC_IS_READY_RETRY_DELAY_SEC.  Having them default to None doesn't make sense because the code below would blow up with those defaults.

DhuldevValekar3 replied:
Yes, Addressed.

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+    :param retries: The number of times to check if the iDRAC is
+                    ready. If None, the value of ready_retries that
+                    was provided when the object was created is
+                    used.
```
Need to remove "If None ... used." above since that's not correct for this implementation.

DhuldevValekar3 replied:
Done, Removed it.

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
+    :param retry_delay: The number of seconds to wait between
+                        retries. If None, the value of
+                        ready_retry_delay that was provided when the
+                        object was created is used.
```
Need to remove "If None ... used." above since that's not correct for this implementation.

DhuldevValekar3 replied:
Done, Removed it.

cdearborn commented on ironic/drivers/modules/redfish/utils.py:
```diff
@@ -274,3 +278,100 @@ def _get_system():
                       'node %(node)s. Error: %(error)s',
                       {'address': driver_info['address'],
                        'node': node.uuid, 'error': e})
```
These changes are very iDRAC specific, so I think we should create ironic/ironic/drivers/modules/drac/utils.py and move them there.

DhuldevValekar3 replied:
Hi Chris, above method _get_system() in redfish/utils.py is come with redfish code only.  I have moved our methods which are specific to idrac in drac/utils.py file.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+                          "for %(node)s, Error : %(error)s" %
+                          {'node':task.node.uuid, 'error':e})
+        url = oem_manager._conn._url
+        redfish_node_ip = re.search(r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}',
```
According to the docs, redfish_address could be a DNS name.  As a result, I think we need to clip out everything between "http[s]://" and the next "/" and use that as the hostname to ping

DhuldevValekar3 replied:
Addressed, used urlparse to clip out hostanme from redfish_address.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +313,173 @@ def set_boot_device(node, device, persistent=False):
+            msg = ("Job queue cleared for node %(node)s via OEM" %
+                    {'node': task.node.uuid})
+            LOG.info(msg)
+            LOG.debug(msg)
```
Need to remove one of the above LOG lines.  Let's remove LOG.debug.  I think it's appropriate to log this at the info level.

DhuldevValekar3 replied:
Done.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +313,173 @@ def set_boot_device(node, device, persistent=False):
+                    "waiting for return to operational state" %
+                    {'node':task.node.uuid})
+            LOG.info(msg)
+            LOG.debug(msg)
```
Let's remove the LOG.debug(msg) line above.  No need to log it twice.

DhuldevValekar3 replied:
Done.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +313,173 @@ def set_boot_device(node, device, persistent=False):
+
+    @task_manager.require_exclusive_lock
+    def set_boot_device(self, task, device, persistent=False):
+        super(
+            DracRedfishManagement,
+            self).set_boot_device(
+            task,
+            device,
+            persistent=False)
```
The call to set_boot_device causes a config job to be created under the covers.  We need to make sure that there are no pending config jobs prior to setting the boot device, otherwise setting the boot device will fail with an error like "a pending config job already exists".
As a result, we need to move this call to super.set_boot_device to after the "if" statement below.  The test to see if this is working is to create a pending config job, and then try running one of these new cleaning steps.  Ironic will call set_boot_device to set the node to PXE boot, and boot up IPA.  After that, it will run the cleaning step.  If the pending job is not deleted then the attempt to set the boot device will fail.

DhuldevValekar3 replied:
Yes Chris, we should move set_boot_device call after deleting pending jobs. Addressed it.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +313,173 @@ def set_boot_device(node, device, persistent=False):
+                                        "%(node)s" %
+                                        {'node':task.node.uuid})
+                            LOG.info(info_msg)
+                            return None
```
We can't return here because we need to execute super.set_boot_device after this code runs.

DhuldevValekar3 replied:
Done, Removed "return None" line.

DhuldevValekar3 commented:
> Hey Dhuldev, this is coming along nicely. Take a look at the new comments I left. One thing is that because the changes to utils are idrac specific, I don't think this will be accepted upstream. We should move that code into a utils.py in the drac directory.
> Another thing is that any place that we log something, we should always include some identifying information, like the IP or GUID for a node, etc.

Hi Chris, 
1. I have moved the our changes to drac/utils.py that are idrac specific. Above comment which you gave given on _get_sysytem() method is not belong to our code, this method came with redfish code.
3. Moved super.set_boot_device() method call after deleting unfinished jobs. 
2. Included identifying information like IP or GUID while logging. 

I have not yet modified unit test cases, once these changes are fine then I will work on unit test

cdearborn commented on ironic/drivers/modules/drac/utils.py:
```diff
@@ -0,0 +1,121 @@
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
```
Can you replace this header with the one we use for JetStream?  The copyright date should be 2020.
See: https://github.com/dsp-jetpack/JetPack/blob/master/src/pilot/customize_image.sh

DhuldevValekar3 replied:
Addressed.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +314,170 @@ def set_boot_device(node, device, persistent=False):
+                LOG.error(error_msg)
+                raise exception.RedfishError(error=error_msg)
+            try:
+                reset_job_response = oem_manager.reset_idrac(task)
```
The above line does a graceful reset.  We need to do a force reset.

DhuldevValekar3 replied:
Done, Added force='Force' parameter to reset_idrac method.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -302,14 +314,170 @@ def set_boot_device(node, device, persistent=False):
+
+                    except sushy.exceptions.SushyError as e:
+                        error_msg = ("Sushy OEM extension Python package "
+                                    "sushy-oem-idrac failed to clear job queue"
```
Can we change the above to "sushy-oem-idrac failed to delete unfinished jobs"

DhuldevValekar3 replied:
Addressed. Added meaningful sentence.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -821,3 +837,105 @@ def test_clear_job_queue(self, mock_get_drac_client):
+        mock_manager = mock.MagicMock()
+        mock_system.managers = [mock_manager]
+        mock_manager_oem = mock_manager.get_oem_extension.return_value
+        mock_manager_oem = mock_manager.get_oem_extension.return_value
```
I think you can delete 1 of the 2 lines above

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -821,3 +837,105 @@ def test_clear_job_queue(self, mock_get_drac_client):
+                                  shared=False) as task:
+            task.node.driver_info['redfish_address'] = '1.2.3.4'
+            return_value = task.driver.management.reset_idrac(task)
+            self.assertIsNone(return_value)
```
we could probably cut out the line above.  The doc for reset_idrac in python-dracclient says that it returns True on success, so validating that it returned None is probably not the best thing to do.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -29,13 +29,29 @@
+
+_SERVICE_ROOT = '/redfish/v1/Managers/'
+
+authenticator = sushy.auth.BasicAuth(
+                    DRAC_REDFISH_INFO_DICT["redfish_username"], 
+                    DRAC_REDFISH_INFO_DICT["redfish_password"])
+url = '%s%s' % (DRAC_REDFISH_INFO_DICT["redfish_address"], _SERVICE_ROOT)
+conn = sushy.Sushy(url, verify=False, auth=authenticator)
+manager = conn.get_manager('iDRAC.Embedded.1')
+oem_manager = manager.get_oem_extension('Dell')
```
Please sync up with Mahendra on how we're going to mock the sushy library.  See Julia's and my comments here: https://review.opendev.org/#/c/693328/7/ironic/tests/unit/drivers/modules/drac/test_inspect.py

DhuldevValekar3 replied:
Thanks Chris, sure I will sync up with Mahendra and will address this.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -821,3 +827,99 @@ def test_clear_job_queue(self, mock_get_drac_client):
 
             self.assertIsNone(return_value)
+
+@mock.patch.object(drac_mgmt, 'redfish_utils', autospec=True)
```
So I think we want to add the following here:
@mock.patch.object(drac_common, 'get_sushy_oem_manager', spec_set=True,
                   autospec=True)

DhuldevValekar3 replied:
Addressed.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -821,3 +827,99 @@ def test_clear_job_queue(self, mock_get_drac_client):
+            for prop in redfish_utils.COMMON_PROPERTIES:
+                self.assertIn(prop, properties)
+
+    def test_clear_job_queue(self, mock_redfish_utils):
```
Then change the method signatures like this:
def test_clear_job_queue(self, mock_get_sushy_oem_manager, mock_redfish_utils):

DhuldevValekar3 replied:
Addressed.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -821,3 +827,99 @@ def test_clear_job_queue(self, mock_get_drac_client):
+
+    def test_clear_job_queue(self, mock_redfish_utils):
+        mock_system = mock_redfish_utils.get_system.return_value
+        mock_manager = mock.MagicMock()
+        mock_system.managers = [mock_manager]
+        mock_manager_oem = mock_manager.get_oem_extension.return_value
```
Then, I don't think you need any of this - just use mock_get_sushy_oem_manager

DhuldevValekar3 replied:
Addressed.

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -324,27 +324,25 @@ def clear_job_queue(self, task):
-                          {'node':task.node.uuid, 'error':e})
-                LOG.error(error_msg)
-                raise exception.RedfishError(error=error_msg)
+        try:
+            oem_manager = drac_common.get_sushy_oem_manager(task.node)
+        except sushy.exceptions.OEMExtensionNotFoundError as e:
+            error_msg = (_("Search for Sushy OEM extension package "
+			"'sushy-oem-idrac' failed for node %(node)s. "
+		        "Ensure it is installed. Error: %(error)s") %
+		        {'node': task.node.uuid, 'error': e})
+            LOG.error(error_msg)
+            raise exception.RedfishError(error=error_msg)
```
Let's move the try/except block into get_sushy_oem_manager().  That way, you can simplify this to just one line here, as well as in the places below where we call get_sushy_oem_manager().

DhuldevValekar3 replied:
Addressed. moved try/except block into get_sushy_oem_manager().

cdearborn commented:
Closing this PR since the corresponding patch has been submitted upstream:
https://review.opendev.org/721593
